lib/configuration.js

```diff
@@ -15,150 +15,430 @@
+ * @memberof Configuration
+ * @private
+ * @function _gatherLocalConfiguration
+ * @returns {Undefined} - does not return anything
  */
-
+Configuration.prototype._gatherLocalConfiguration = function() {
+  if (isPlainObject(this._givenConfiguration)) {
+    if (isBoolean(this._givenConfiguration.reportUncaughtExceptions)) {
+      this._reportUncaughtExceptions = this._givenConfiguration
+        .reportUncaughtExceptions;
+    }
+    if (isString(this._givenConfiguration.key)) {
+      this._key = this._givenConfiguration.key;
+    }
+  }
+};
 /**
- * Orchestrates the gathering of configuration based on submission of the
- * initConfiguration parameter. If all optional properties are given in the
- * initConfiguration parameter then the only environmental variable that will
- * be read is the NODE_ENV variable. If initConfiguration is not supplied or
- * only partially supplied then gatherConfiguration will attempt to read
- * environmental variables to supplement information as necessary.
- * @function gatherConfiguration
- * @param {ConfigurationOptions} initConfiguration - the options object
- * @returns {NormalizedConfigurationVariables} - a configuration object
- *  supplemented with environmental information as necessary
+ * The callback for `utils.getProjectNumber` this function is responsible for
+ * determining whether or not an error occured in the metadata transaction and,
+ * if one did not, assigning the value of the transaction (the project number)
+ * to the instance property `_projectNumber`.
+ * @memberof Configuration
+ * @private
+ * @function _assimilateProjectNumber
+ * @param {Error|Undefined} err - If applicable, the error that occured during 
+ *  the transaction
+ * @param {Null|String} projectNum - The project number if successfully 
+ *  retrieved, null otherwise.
+ * @returns {Undefined} - does not return anything
  */
-function gatherConfiguration(initConfiguration) {
-
-  var projectId = null;
-  var reportUncaughtExceptions = true;
-  var serviceContext = null;
-  var key = null;
-  var shouldReportErrorsToAPI = false;
-
-  if (env.NODE_ENV === 'production') {
-
-    shouldReportErrorsToAPI = true;
+Configuration.prototype._assimilateProjectNumber = function(err, projectNum) {
+  if (isNull(err) && isString(projectNum)) {
+    this._projectNumber = projectNum;
   } else {
-
-    console.log([
-      'Warning: the NODE_ENV environmental variable is not set to',
-      '\'production\'. To send errors to the Stackdriver Error Reporting API',
-      'please set the NODE_ENV to \'production\'.'
-    ].join(' '));
-  }
-
-  if (isPlainObject(initConfiguration)) {
-
-    if (has(initConfiguration, 'projectId')) {
-
-      projectId = initConfiguration.projectId;
-    }
-
-    if (has(initConfiguration, 'reportUncaughtExceptions')) {
-
-      reportUncaughtExceptions = !!initConfiguration.reportUncaughtExceptions;
+    if (this._logger) {
+      this._logger.warn(
```

Comment:
Okay this ends up working because the trace agent seems to be using local config first before checking with the metadata service. (/cc @matthewloring to reconcile the ordering). Regardless the warning should only show up when the user has done something wrong.